### PR TITLE
E2e fixes

### PR DESCRIPTION
**Pull request by riazmoola**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -4,22 +4,56 @@
...
+    "expect"
   ],
-  "keywords": ["ssh","sshclient","remote","cli","net","expect"],
   "main": "lib/index.js",
-  "engines": { "node": ">= 0.2.0" }
+  "engines": {
+    "node": ">= 0.2.0"
+  },
+  "dependencies": {
+    "mocha": "^5.0.4",
```
Comment (jbernalvallejo) [inline package.json]: Mocha is a test runner, you probably want it in devDependencies.

File: promiseWrapper.js [hunk 2]
```diff
@@ -1,138 +1,77 @@
...
+
+
+  // execCommand(command: string, options: { cwd: string, stdin: string } = {}): Promise<{ stdout: string}>
+  execCommand(value, options) {
     return new Promise((resolve, reject) => {
-      console.log('in promise')
+      if (options.cwd) {
+        value = `cd ${shellEscape([options.cwd])} 1> /dev/null 2> /dev/null; ${value}`;
+      }
+      var wholeLine = '';
```
Comment (jbernalvallejo) [inline promiseWrapper.js]: var is deprecated. We should use let or const.

https://odetocode.com/blogs/scott/archive/2014/07/31/the-features-of-es6-part-1-let.aspx
https://odetocode.com/blogs/scott/archive/2014/08/06/features-of-es6-part-2-const.aspx

**Review by jbernalvallejo — APPROVED** [review]
Looks good Riaz!

File: promiseWrapper.js (hunk 2 above)
Comment (riazmoola) [reply promiseWrapper.js]: ok cool

File: package.json (hunk 1 above)
Comment (riazmoola) [reply package.json]: done

File: README.md [hunk 3]
```diff
@@ -37,8 +37,36 @@ Download and install dependencies
 
 More samples at `samples/` directory.
 
+Alternatively use the promiseWrapper
+
+    var options = {
```
Comment (jbernalvallejo) [inline README.md]: Remember, var is deprecated ;)

File: README.md [hunk 4]
```diff
@@ -37,8 +37,36 @@ Download and install dependencies
...
+        return mySSH.execCommand('ls', options);
+      })
+      .then((result) => {
+        console.log('contents of /user/ is', result.stdout)
+        return mySSH.dispose();
+      }).then((string) => {
+        console.log('SSH session has been closed successfully')
+        done();
+      })
+      .catch(done);
```
Comment (jbernalvallejo) [inline README.md:65]: Good example!

File: README.md [hunk 5]
```diff
@@ -173,6 +201,57 @@ Sample:
...
+    
+#### dispose
+Dispose of the connection. Can reconnect by calling connect();
+
+Sample:
+    
+    mySSH.dispose().then((result) => {
+        console.log('Attempting to close SSH session\n', result)
+    });
+
```
Comment (jbernalvallejo) [inline README.md]: Good documentation and examples, well done!

File: promiseWrapper.js [hunk 6]
```diff
@@ -1,138 +1,107 @@
...
+          wholeLine += data;
+        } else {
+          if (lastChunk === true) {
+            let res = '';
+            this.connection.removeAllListeners('output');
+            if (wholeLine.indexOf(value) !== -1) {
+              debug('$', value);
+              res = wholeLine.replace(value, '');
+            } else {
+              console.error(`Expected ${value} to be present in ${wholeLine}`);
```
Comment (jbernalvallejo) [inline promiseWrapper.js]: We could still debug to log this error message. However, with this approach an error will get logged regardless of the DEBUG content so I reckon this is a good approach.

**Review by jbernalvallejo — COMMENTED** [review]
Well done Riaz! It looks very good. I still have some comments about general structure of the project:
- As far as I can see, promiseWrapper is just a 'promisified' envelope around the ssh client. Hence, I think this wrapper should live in the lib folder as the main index.js does.
- Also, this library is going to be exporting two ssh clients: old one and the new promisified one. I will review how we are exporting these so it is more intuitive. For ie, I may have a index.js in the root of the project which simply exposes both clients. Give it some thought.

File: index.js [hunk 7]
```diff
@@ -0,0 +1,8 @@
+'use strict';
+const promiseWrapper = require('./lib/promiseWrapper');
+const NodeSSH = require('./lib/NodeSSH');
+
+module.exports = {
+  promiseWrapper: promiseWrapper,
+  NodeSSH: NodeSSH
```
Comment (jbernalvallejo) [inline index.js]: So with this config, we would need to require it this way:

`require('NodeSSH').NodeSSH`
or
`require('NodeSSH').promiseWrapper`

I wonder if we could do something more elegant where `require('NodeSSH')` returns the default client based on callbacks (this way you don't break current contract). However, this client could have a function called `prosimify` that returns the one with the promise wrapper around it. This way when consuming it:

```
require('NodeSSH'); // we are good to go with the default client, same API as before so current consumers don't break
require('NodeSSH').promisify(); // what promiseWrapper returns
```

What do you think?

**Comment by riazmoola** [comment]
Oh I didn't think of that. Would be better yeah.
________________________________

**Pull request merged**